pkg/api/mlflow/dao/repositories/namespace_cached.go

```diff
@@ -0,0 +1,186 @@
+) (*NamespaceCachedRepository, error) {
+	cache, err := lru.New[string, *models.Namespace](1000)
+	if err != nil {
+		return nil, eris.Wrap(err, "error creating lru cache for namespace entities")
+	}
+
+	// pre load all namespaces into cache.
+	var namespaces []models.Namespace
+	if err := db.Find(&namespaces).Error; err != nil {
+		return nil, eris.Wrapf(err, "error getting namespaces")
+	}
+	//nolint:gosec
+	for _, namespace := range namespaces {
+		cache.Add(namespace.Code, &namespace)
+	}
+
+	repository := NamespaceCachedRepository{
+		db:                  db,
+		cache:               cache,
+		listener:            listener,
+		namespaceRepository: namespaceRepository,
+	}
+
+	go func() {
+		for data := range listener.Listen() {
+			if err := repository.processEvent(data); err != nil {
+				log.Errorf(`error processing incoming event: %s, error: %+v`, data, err)
+			}
+		}
+	}()
+	return &repository, nil
+}
+
+// Create creates new models.Namespace entity.
+func (r NamespaceCachedRepository) Create(ctx context.Context, namespace *models.Namespace) error {
+	if err := r.namespaceRepository.Create(ctx, namespace); err != nil {
+		return eris.Wrap(err, "error creating cached namespace entity")
+	}
+
+	// trigger database event to notify current instance and
+	// other instances to create record in theirs local cache.
+	if err := r.sendEvent(NamespaceEventActionCreated, namespace); err != nil {
+		return eris.Wrap(err, "error sending database event")
+	}
+	return nil
+}
+
+// Update updates existing models.Namespace entity.
+func (r NamespaceCachedRepository) Update(ctx context.Context, namespace *models.Namespace) error {
+	if err := r.namespaceRepository.Update(ctx, namespace); err != nil {
+		return eris.Wrap(err, "error updating cached namespace entity")
+	}
+
+	// trigger database event to notify current instance and
+	// other instances to update record in theirs local cache.
+	if err := r.sendEvent(NamespaceEventActionUpdated, namespace); err != nil {
+		return eris.Wrap(err, "error sending database event")
+	}
+	return nil
+}
+
+// GetByCode returns namespace by its Code.
+func (r NamespaceCachedRepository) GetByCode(ctx context.Context, code string) (*models.Namespace, error) {
+	result, ok := r.cache.Get(code)
+	if ok {
+		return result, nil
+	}
+
+	namespace, err := r.namespaceRepository.GetByCode(ctx, code)
+	if err != nil {
+		return nil, eris.Wrapf(err, "error getting cached namespace by code: %s", code)
+	}
+
+	// trigger database event to notify current instance and
+	// other instances to add record to theirs local cache.
+	if err := r.sendEvent(NamespaceEventActionFetched, namespace); err != nil {
+		return nil, eris.Wrap(err, "error sending database event")
+	}
+	return namespace, nil
+}
+
+// GetByID returns namespace by its ID.
+func (r NamespaceCachedRepository) GetByID(ctx context.Context, id uint) (*models.Namespace, error) {
+	return r.namespaceRepository.GetByID(ctx, id)
```

Comment:
not checking cache here?

Reply:
nope unfortunately, because:
- GetByID used mostly in UI.
- GetByCode used in middleware, so it means that each call to the service will call this method so we have to cache it.
- we store `namespace` in `cache` using `namespace` id.
actually right now I think when we call `GetByID` we can store result inside the cache.